test/e2e/gardener/shoot/shoot.go

```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
+			"app":              "machine",
+			"machine-provider": "local",
+		})).Should(Succeed())
+
+		for _, item := range beforeStartMachinePodList.Items {
+			beforeStartMachinePodNames.Insert(item.Name)
+		}
+	}, SpecTimeout(time.Minute))
+
+	return beforeStartMachinePodNames
+}
+
+// ItShouldCompareMachinePodNamesAfter compares the machine pod names before and after running the required tests.
+func ItShouldCompareMachinePodNamesAfter(s *ShootContext, beforeStartMachinePodNames sets.Set[string]) {
+	GinkgoHelper()
+
+	It("Compare machine pod names", func(ctx SpecContext) {
+		afterStartMachinePodList := &corev1.PodList{}
+		Eventually(ctx, s.SeedKomega.List(afterStartMachinePodList, client.InNamespace(s.Shoot.Status.TechnicalID), client.MatchingLabels{
+			"app":              "machine",
+			"machine-provider": "local",
+		})).Should(Succeed())
+
+		afterStartMachinePodNames := sets.New[string]()
+		for _, item := range afterStartMachinePodList.Items {
+			afterStartMachinePodNames.Insert(item.Name)
+		}
+
+		Expect(beforeStartMachinePodNames.UnsortedList()).To(ConsistOf(afterStartMachinePodNames.UnsortedList()))
+	}, SpecTimeout(time.Minute))
+}
+
+// ItShouldRewriteOsRelease rewrites the /etc/os-release file for all machine pods to ensure that the version is overwritten for tests.
+// This is a workaround for the fact that the machine image version is not available in the os-release file in the local provider.
+func ItShouldRewriteOsRelease(s *ShootContext) {
+	GinkgoHelper()
+
+	It("should rewrite the /etc/os-release for all machines", func(ctx SpecContext) {
+		podList := &corev1.PodList{}
+		Eventually(ctx, s.SeedKomega.List(podList,
+			client.InNamespace(s.Shoot.Status.TechnicalID),
+			client.MatchingLabels{
+				"app":              "machine",
+				"machine-provider": "local",
+			},
+		)).Should(Succeed())
+
+		for _, pod := range podList.Items {
+			node := &corev1.Node{}
+			Eventually(func() error {
+				return s.ShootClient.Get(ctx, client.ObjectKey{Namespace: pod.Namespace, Name: pod.Name}, node)
+			}).Should(Succeed())
+
+			Expect(node.Labels).To(HaveKey(v1beta1constants.LabelWorkerPool))
+
+			poolIndex := slices.IndexFunc(s.Shoot.Spec.Provider.Workers, func(pool gardencorev1beta1.Worker) bool {
+				return pool.Name == node.Labels[v1beta1constants.LabelWorkerPool]
+			})
+			Expect(poolIndex).To(BeNumerically(">", -1))
+
+			_, _, err := s.SeedClientSet.PodExecutor().Execute(ctx,
+				pod.Namespace,
+				pod.Name,
+				local.MachinePodContainerName,
+				"sed",
+				"-i", "-E",
+				fmt.Sprintf(
+					`s/^PRETTY_NAME="[^"]*"/PRETTY_NAME="Machine Image Version %s (version overwritten for tests, check VERSION_ID for actual version)"/`,
+					*s.Shoot.Spec.Provider.Workers[poolIndex].Machine.Image.Version,
+				),
+				"/etc/os-release",
+			)
+
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}, SpecTimeout(2*time.Minute))
```

Comment:
This is weird. Shouldn't this be done by `provider-local` itself somehow? Doing this in the e2e test feels wrong.

Reply:
This is supposed to be done by the user in the real scenario, here the e2e test is the "user".
We can also write a controller in provider-local to watch for machines and do this, but this is not needed in any other test other than in-place.

Reply:
Fixed with https://github.com/gardener/gardener/pull/11953/commits/03b9f2aeb2babb312d45b9618cf333f01e4996d5